### PR TITLE
Fix job-master leak memory when submitting distributed jobs

### PR DESCRIPTION
**Pull request by liiuzq-xiaobai**
fix:fix job-master leak memory when submitting a large number of distributed jobs(DIST_LOAD/DIST_CP/Persist jobs)

### What changes are proposed in this pull request?

Start a periodic thread to clear expired jobs information that cannot be trace by the client in CmdJobTracker.The default retention time is 1day,which is the same configuration as LoadV2.

### Why are the changes needed?

When many jobs are submitted,the job master finally will have an oom problem, we can find that the cmdJobTracker retains the residual job information and not cleaned regularly, resulting in memory leaks.

### Does this PR introduce any user facing changes?

Please list the user-facing changes introduced by your change, including
1.add Configuration:
          alluxio.job.master.job.trace.retention.time=xx,the default value is 1d.

Related issue: #18635

### CODE REVIEW

File: job/server/src/main/java/alluxio/master/job/tracker/CmdJobTracker.java [hunk 1]
```diff
@@ -270,4 +289,29 @@ public CmdStatusBlock getCmdStatusBlock(long jobControlId)
             .collect(Collectors.toList());
     return new CmdStatusBlock(cmdInfo.getJobControlId(), blockList, cmdInfo.getOperationType());
   }
+
+  private void cleanUnableTracedJobs() {
+    long currentTime = System.currentTimeMillis();
+    for (Map.Entry<Long, CmdInfo> x : mInfoMap.entrySet()) {
+      CmdInfo cmdInfo = x.getValue();
+      if (currentTime - cmdInfo.getJobSubmissionTime() > sTraceRetentionTime) {
```
Comment (ccy00808) [inline job/server/src/main/java/alluxio/master/job/tracker/CmdJobTracker.java:297]: if (currentTime - cmdInfo.getJobSubmissionTime() > sTraceRetentionTime)

(currentTime - cmdInfo.getJobFinishedTime() - sTraceRetentionTime = 1s

Users will never be able to query the results of this load？

**Comment by ccy00808** [comment]
image

File: job/server/src/main/java/alluxio/master/job/tracker/CmdJobTracker.java [hunk 2]
```diff
@@ -270,4 +289,29 @@ public CmdStatusBlock getCmdStatusBlock(long jobControlId)
             .collect(Collectors.toList());
     return new CmdStatusBlock(cmdInfo.getJobControlId(), blockList, cmdInfo.getOperationType());
   }
+
+  private void cleanUnableTracedJobs() {
+    long currentTime = System.currentTimeMillis();
+    for (Map.Entry<Long, CmdInfo> x : mInfoMap.entrySet()) {
```
Comment (ccy00808) [inline job/server/src/main/java/alluxio/master/job/tracker/CmdJobTracker.java:295]: Will the mCoordinators in the PlanTracker class also occupy a lot of memory? If the capacity is set too small, will it cause abnormal results when getting status?

  https://github.com/Alluxio/alluxio/blob/8dc6653ddb86b92be6c4ea1cffe0d963e1c37d7c/job/server/src/main/java/alluxio/master/job/JobMaster.java#L580

File: job/server/src/main/java/alluxio/master/job/tracker/CmdJobTracker.java [hunk 3]
```diff
@@ -270,4 +289,29 @@ public CmdStatusBlock getCmdStatusBlock(long jobControlId)
...
+      CmdInfo cmdInfo = x.getValue();
+      if (currentTime - cmdInfo.getJobSubmissionTime() > sTraceRetentionTime) {
+        try {
+          Status jobStatus = getCmdStatus(cmdInfo.getJobControlId());
+          if (jobStatus.isFinished()) {
+            mInfoMap.remove(cmdInfo.getJobControlId());
+            LOG.debug("JobControlId:{} has been cleaned in CmdJobTracker,"
```
Comment (ccy00808) [inline job/server/src/main/java/alluxio/master/job/tracker/CmdJobTracker.java:302]: Debug change to Info and add path info?

File: job/server/src/main/java/alluxio/master/job/tracker/CmdJobTracker.java [hunk 4]
```diff
@@ -270,4 +289,29 @@ public CmdStatusBlock getCmdStatusBlock(long jobControlId)
...
+
+  private void cleanUnableTracedJobs() {
+    long currentTime = System.currentTimeMillis();
+    for (Map.Entry<Long, CmdInfo> x : mInfoMap.entrySet()) {
+      CmdInfo cmdInfo = x.getValue();
+      if (currentTime - cmdInfo.getJobSubmissionTime() > sTraceRetentionTime) {
+        try {
+          Status jobStatus = getCmdStatus(cmdInfo.getJobControlId());
+          if (jobStatus.isFinished()) {
+            mInfoMap.remove(cmdInfo.getJobControlId());
```
Comment (ccy00808) [inline job/server/src/main/java/alluxio/master/job/tracker/CmdJobTracker.java:301]: Can add extend callback to remove all PlanCoordinators at the cmdInfo?

File: job/server/src/main/java/alluxio/master/job/tracker/CmdJobTracker.java (hunk 1 above)
Comment (ccy00808) [reply job/server/src/main/java/alluxio/master/job/tracker/CmdJobTracker.java:307]: [TODO]

File: job/server/src/main/java/alluxio/master/job/plan/PlanTracker.java [hunk 5]
```diff
@@ -300,6 +173,17 @@ public Set<Long> findJobs(String name, List<Status> statusList) {
...
   }
 
+  /**
+   * Remove expired jobs in PlanTracker.
+   * @param jobIds the list of removed jobId
+   */
+  public void removeJobs(List<Long> jobIds) {
+    mWorkflowTracker.cleanup(jobIds);
+    for (Long jobId : jobIds) {
+      mCoordinators.remove(jobId);
```
Comment (ccy00808) [inline job/server/src/main/java/alluxio/master/job/plan/PlanTracker.java:183]: Does the mFailed collection also need to be cleaned up?

File: job/server/src/main/java/alluxio/master/job/plan/PlanTracker.java [hunk 6]
```diff
@@ -178,84 +124,17 @@ public PlanCoordinator getCoordinator(long jobId) {
    * @throws ResourceExhaustedException if there is no more space available in the job master
    */
   public synchronized void run(PlanConfig jobConfig, CommandManager manager,
-      JobServerContext ctx, List<WorkerInfo> workers, long jobId) throws
+                               JobServerContext ctx, List<WorkerInfo> workers, long jobId) throws
       JobDoesNotExistException, ResourceExhaustedException {
     checkActiveSetReplicaJobs(jobConfig);
-    if (removeFinished()) {
+    if (mCoordinators.size() < mCapacity) {
```
Comment (ccy00808) [inline job/server/src/main/java/alluxio/master/job/plan/PlanTracker.java:130]: Why is there the capacity limit. the running job limits is here:
https://github.com/Alluxio/alluxio/blob/1359cc53018413e28044152a24d0c762b300f25b/job/server/src/main/java/alluxio/master/job/tracker/DistLoadCliRunner.java#L159

Comment (liiuzq-xiaobai) [reply job/server/src/main/java/alluxio/master/job/plan/PlanTracker.java:130]: Compatibility with some unit tests.

**Comment by ccy00808** [comment]
@jja725 Can help us review the PR?

**Review by jja725 — APPROVED** [review]
LGTM, thanks for the improvement!

**Review by ccy00808 — APPROVED** [review]

**Comment by ccy00808** [comment]
alluxio-bot, merge this please.

**Comment by ccy00808** [comment]
alluxio-bot, cherry-pick this to branch-2.10 please

**Pull request merged**